Algorithms/Problems/CoinChange/DynamicCoinChangeSolver.cs

```diff
@@ -0,0 +1,179 @@
+                dict[currentCoin] = changeArray;
+                currentCoin++;
+            }
+
+            return dict;
+        }
+
+        /// <summary>
+        /// Gets a next coin value, such that changes array contains the minimal change overall possible changes.
+        /// For example, having coin N = 6 and A = [1,3,4] coins array.
+        /// The minimum next coin for 6 will be 3, because changes of 3 by A = [1,3,4] contains 0, the minimal change.
+        /// </summary>
+        /// <param name="coin">Coin to be exchanged.</param>
+        /// <param name="exchanges">Dictionary of exchanges for [1, coin].</param>
+        /// <returns>Index of the next coin with minimal exchange.</returns>
+        public static int GetMinimalNextCoin(int coin, Dictionary<int, int[]> exchanges)
+        {
+            var nextCoin = int.MaxValue;
+            var minChange = int.MaxValue;
+
+            var coinChanges = exchanges[coin];
+
+            foreach (var change in coinChanges)
+            {
+                if (change == 0)
+                {
+                    return 0;
+                }
+
+                var currentChange = exchanges[change];
+                var min = currentChange.Min();
+
+                var minIsLesser = min < minChange;
+
+                if (minIsLesser)
+                {
+                    nextCoin = change;
+                    minChange = min;
+                }
+            }
+
+            return nextCoin;
+        }
+
+        /// <summary>
+        /// Performs a coin change such that an amount of coins is minimal.
+        /// </summary>
+        /// <param name="coin">The value of coin to be exchanged.</param>
+        /// <param name="coins">An array of available coins.</param>
+        /// <returns>Array of exchanges.</returns>
+        public static int[] MakeCoinChangeDynamic(int coin, int[] coins)
+        {
+            var changesTable = GenerateChangesDictionary(coin, coins);
+            var list = new List<int>();
+
+            var currentCoin = coin;
+            var nextCoin = int.MaxValue;
+
+            while (nextCoin != 0)
+            {
+                nextCoin = GetMinimalNextCoin(currentCoin, changesTable);
+                var difference = currentCoin - nextCoin;
+                list.Add(difference);
+                currentCoin = nextCoin;
+            }
+
+            var result = list.ToArray();
+
+            return result;
+        }
+
+        private static void ValidateCoin(int coin)
+        {
+            if (coin <= 0)
+            {
+                throw new InvalidOperationException($"The coin cannot be lesser or equal to zero {nameof(coin)}.");
+            }
+        }
+
+        private static void ValidateCoinsArray(int[] coinsArray)
+        {
+            if (coinsArray == null)
+            {
+                throw new ArgumentNullException(nameof(coinsArray));
+            }
```

Comment:
Please remove this check, int[] is not nullable. I get that the user can technically still pass null there, but it's their problem. The interface states clearly that coins are not nullable.